File: 0885-spiral-matrix-iii/0885-spiral-matrix-iii.py
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        unordered_map<char, array<int, 4>> dir {
            {'R', {1, 0, 0, 1}},
            {'D', {0, -1, 1, 0}},
            {'L', {-1, 0, 0, -1}},
            {'U', {0, 1, -1, 0}},
        };
        
        function<pair<int,int>(int, int, char)> move;
        function<bool(int, int)> inbound = [&](int r, int c) {
            return (r >= 0 and r < rows and c >= 0 and c < cols);
        };
        vector<vector<int>> ans;
        ans.push_back({rStart, cStart});
        set<pair<int,int>> vis;
        vis.insert({rStart, cStart});
        cStart++;
        
        move = [&](int row, int col, char direc) {
            auto [r, c, dr, dc] = dir[direc];
            while (vis.count({row + r, col + c})) {
                if (inbound(row, col))
                    ans.push_back({row, col});
                vis.insert({row, col});
                row += dr;
                col += dc;
            }
            if (vis.count({row + r, col + c}))
                if (inbound(row, col))
                    ans.push_back({row, col});
            vis.insert({row, col});
            pair<int,int> cur = {row, col};
            return cur;
        };
        int r = rStart, c = cStart;
        while (ans.size() < rows * cols) {
            tie(r, c) = move(r, c, 'R');
            tie(r, c) = move(r, c, 'D');
            tie(r, c) = move(r, c, 'L');
            tie(r, c) = move(r, c, 'U');
        }
        return ans;
    }
};

File: 0885-spiral-matrix-iii/0885-spiral-matrix-iii.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        int dir[][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        vector<vector<int>> ans;
        int row = rStart, col = cStart;
        for (int step = 1, d = 0; ans.size() < rows * cols; step++) {
            for (int i = 0; i < 2; i++) {
                for (int j = 0; j < step; j++) {
                    if (row >= 0 and row < rows and col >= 0 and col < cols)
                        ans.push_back({row, col});
                    row += dir[d][0];
                    col += dir[d][1];
                }
                d = (d + 1) % 4;
            }
        }
        return ans;
    }
};